### PR TITLE
npm-upgrade

### PR DESCRIPTION
**Pull request by rvilarl**
Npm packages updated

### CODE REVIEW

**Comment by sschmidTU** [comment]
Cross-blob 3 can't be used with `require`, so at the very least this would necessitate merging the generateImages script to be an ESModule. We also need to check other effects of these upgrades.

**Comment by rvilarl** [comment]
Simon how should I run the visual regression test?

**Comment by sschmidTU** [comment]
As I said, you can't run it on this PR currently, because it needs `generateImages_browserless.js`, which doesn't work with cross-blob 3, because it's an ESModule now, so you can't `require` it. So it looks like the generateImages script needs to be migrated to be an ESModule and use imports instead of `require`s.

**Comment by sschmidTU** [comment]
Here's how you normally do it:
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/wiki/Testing

**Comment by rvilarl** [comment]
Migrated!

```
Found 117 current and 117 blessed png files (not tested if valid). Continuing.\n
Running 117 tests with threshold 0.0001 (nproc=8)...
Progress : [########################################] 100%
Results stored in ./visual_regression/diff/results.txt
All images with a difference over threshold, 0.0001, are
available in ./visual_regression/diff, sorted by perceptual hash.

Success - All diffs under threshold!
```

File: test/Util/DiffImages_Test_Experimental.ts [hunk 1]
```diff
@@ -58,6 +58,6 @@ describe("GeneratePNGImages", () => {
             //fs.writeFileSync(fileName, imageBuffer, { encoding: "base64" });
 
             done();
-            }).timeout(10000);
+            }).timeout(30000);
```
Comment (sschmidTU) [inline test/Util/DiffImages_Test_Experimental.ts:61]: this seems to be an excessive timeout, did you ever exceed 10s? how about 20s?

**Comment by sschmidTU** [comment]
> Migrated!

Awesome, thank you very much! I'll take a look soon.

File: .appveyor.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 image: Visual Studio 2017
 environment:
-  timeout: 10000
+  timeout: 30000
```
Comment (sschmidTU) [inline .appveyor.yml:3]: please also reduce this, if possible to 15s, that should be enough. there are some tests that deliberately go to timeout, and 20 seconds more on a lot of tests and builds is a lot in total.

File: .travis.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ node_js:
 # - '10' # fails on Travis since upgrading to Vexflow 1.2.90 (still passes on AppVeyor), for Mxl_Test. Node 12 works.
 - '12'
 env:
-  - timeout=10000
+  - timeout=30000
```
Comment (sschmidTU) [inline .travis.yml:8]: we don't use travis currently, so there's not really a need to adjust this.

**Comment by sschmidTU** [comment]
@rvilarl The Create PDF function (button) in the demo is broken.
console:
```console
Uncaught TypeError: _node_modules_svg2pdf_js_dist_svg2pdf_umd_min__WEBPACK_IMPORTED_MODULE_3__ is not a function
    at createPdf (demo.js:899)
    at HTMLDivElement.printPdfBtn.onclick (demo.js:412)
```
line in the built demo (`npm start`):
`            _node_modules_svg2pdf_js_dist_svg2pdf_umd_min__WEBPACK_IMPORTED_MODULE_3__(svgElement, pdf, {`

**Comment by sschmidTU** [comment]
Using `svg2pdf.svg2pdf(svgElement, pdf, {` in `index.js:870` at least download a PDF, but returns an empty pdf and throws a new error in jspdf:
`jspdf.min.js:40 Uncaught (in promise) TypeError: Cannot redefine property: sx`

**Comment by sschmidTU** [comment]
@rvilarl I fixed the pdf function, we can even remove one deprecated import now.
Please apply this patch:
[pdf_patch.txt](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/7611889/pdf_patch.txt)

**Comment by rvilarl** [comment]
@sschmidTU how do I apply the patch ?

**Comment by sschmidTU** [comment]
@rvilarl git apply patchfile

**Comment by rvilarl** [comment]
Done

**Comment by sschmidTU** [comment]
@rvilarl `npm test` isn't working for me locally, is it for you?
karma-server throws errors, starting with:
```console
[karma-server]: Error during file loading or preprocessing
TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at Hash.update (node:internal/crypto/hash:105:11)
    at Object.sha1 (/Users/simon/OSMD/opensheetmusicdisplay-working-copy/node_modules/karma/lib/utils/crypto-utils.js:9:8)
    at runProcessors (/Users/simon/OSMD/opensheetmusicdisplay-working-copy/node_modules/karma/lib/preprocessor.js:70:26)
```

I upgraded to node 16, didn't fix it.

**Comment by rvilarl** [comment]
@sschmidTU yes it works:
```
Finished in 7.976 secs / 5.984 secs @ 16:07:47 GMT+0100 (hora estándar de Europa central)

SUMMARY:
✔ 170 tests completed
ℹ 1 test skipped
```

**Comment by rvilarl** [comment]
@sschmidTU I suggest that you delete the `./node_modules` directory and call `npm install`.

**Comment by sschmidTU** [comment]
@rvilarl I did that twice now, it still doesn't work. Strange.

**Comment by rvilarl** [comment]
@sschmidTU how can we compare our environments?

**Comment by sschmidTU** [comment]
I don't have ng, where is that from?
```console
ng --version
bash: ng: command not found
```

I'm using:
Node: 16.13.0
Package Manager: npm 8.1.0
OS: MacOS

I doubt that Node 14 will make it work.
I also tried it with Node 12 before I upgraded.

**Comment by sschmidTU** [comment]
Still same error with your node and npm versions installed.

**Comment by sschmidTU** [comment]
Okay, I made it run with some hacks, seems to be a problem with karma-webpack 5 and MacOS.
It looks like karma-webpack 5 is not mature yet. You need it for Webpack 5 though.

In `node_modules_karma-webpack/lib/karma-webpack/preprocessor.js`,
I had to replace line 96 with this:
`return bundleContent;`

And in `node_modules_karma-webpack/lib/webpack/plugin.js`, I had to replace line 18 with this:
`webpackFileObj.name.replace("Users", "../Users")`
There is a `../` missing from the path to make it valid.

I guess I'll have to open some issues at karma-webpack...

**Comment by sschmidTU** [comment]
The karma-webpack issue is fixed on their master, just not released yet.
I opened an issue about it here:
https://github.com/ryanclark/karma-webpack/issues/518

**Comment by rvilarl** [comment]
Very good! I am happy that you found it! So perhaps we wait for the fix to come and then merge this one. 
Are you happy with the changes? Should I look into something else?

**Comment by sschmidTU** [comment]
The changes in this PR look good now except for the karma-webpack issue, I'll merge it once that is fixed. (If they don't release a new version in a few days, I'll create the `KarmaWebpackPatch` prebuild system)

**Comment by rvilarl** [comment]
@sschmidTU I just rebased the latest from development. It does not seem that karma-webpack is planning to release in short term. Should we go for the patch?

**Comment by sschmidTU** [comment]
Yeah, probably.

**Comment by rvilarl** [comment]
will you send me a patch to apply? It would be good to merge to avoid continuous rebase. You can also close #1059 with this one.

**Comment by rvilarl** [comment]
@sschmidTU are you still thinking about producing a patch for karma or rather wait? Should I continue to rebase this PR to keep it up to date? I just did by the way

**Comment by sschmidTU** [comment]
Nice. I think I'll make the karma patch now, doesn't look like their release will be anytime soon.

**Comment by sschmidTU** [comment]
Oh well, npm test doesn't work on Windows now, with or without patch... might have to revert this.

**Comment by sschmidTU** [comment]
Karma doesn't even work with the latest karma-webpack master branch, on Windows
So unfortunately, I'm going to have to revert this npm upgrade.

**Comment by sschmidTU** [comment]
For future reference, here's the patch for for adding KarmaWebpackPatch:
[karmaWebpackPatch.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/7729674/karmaWebpackPatch.zip)

This fixes karma-webpack 5.0 on MacOS, but not Windows, unfortunately.

**Comment by sschmidTU** [comment]
I'll keep this on a new branch for now, feat/webpack5.

**Comment by rvilarl** [comment]
It does work in linux. It is a pitty that you had to revert it :(

**Comment by sschmidTU** [comment]
Yeah, it's unfortunate that file paths are still broken for karma-webpack on Windows.
I'll try to see if I can patch it..

**Comment by rvilarl** [comment]
I just run it on windows, what is the error that you get?

**Comment by rvilarl** [comment]
I run "npm test:visual:build" and I just got an error at the end complaining about bash

**Comment by sschmidTU** [comment]
It's not about visual tests, but npm test.
`Error: EINVAL: invalid argument, mkdir 'C:\Users\Simon\AppData\Local\Temp\_karma_webpack_82360\D:\code\osmd\opensheetmusicdisplay\dist\src'`

Maybe this is the problem:
https://github.com/ryanclark/karma-webpack/pull/433

Maybe you don't have different partitions on your hard drive.

**Comment by rvilarl** [comment]
I will give it a try

**Comment by sschmidTU** [comment]
Yes, it's a problem. Visual Regression tests run on Windows using WSL.

**Comment by rvilarl** [comment]
In C:/ it runs well. I believe that this is a reasonalbe limitation that we could state in the README for a while until it is fixed. Don´t you agree?

**Comment by sschmidTU** [comment]
No.

**Comment by rvilarl** [comment]
OK let us try to progress on another front :) Can you help me with the VexFlowPatch issues that I have opened in VewFlow ?

**Comment by sschmidTU** [comment]
Yes, when I have the time.

**Comment by sschmidTU** [comment]
Commented on all the issues.
https://github.com/0xfe/vexflow/issues?q=is%3Aissue+OSMD+VexflowPatch

**Comment by sschmidTU** [comment]
@rvilarl I've decided to merge this to develop soon. I'll try to patch in correct paths for Windows, but if it's too much work, I'll just accept the limitation that on Windows, npm test has to be done on the `C:\` partition.

**Comment by sschmidTU** [comment]
File paths on Windows fixed in f66753d60cf141c9075f1dc5ad47f382b74c7bc8.
Npm test continues working on Linux. MacOS should be the same.

So, now this can be finally merged to develop from feat/webpack5.

@rvilarl For rebasing your PRs etc. you should probably wait until this and the next OSMD Release which is coming soon.

**Comment by rvilarl** [comment]
@sschmidTU what is coming in the next release? I will make a prerelease branch in Vexflow soon and I am also looking on how to fix the Beam issues.

**Comment by sschmidTU** [comment]
@rvilarl Just the commits on develop since the last release, roughly this:
## [1.3.2](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/compare/1.3.1...1.3.2) (2022-01-07 Preview)


### Bug Fixes

* **Cursor:** Fix follow for multiple cursors, can set cursor.cursorOptions.follow for each ([#1111](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1111)) ([37f9002](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/37f9002c5e90a351447854cc926f53ab16107edc))
* **Grace Notes:** Don't draw multiple grace not slashes for a set of grace notes ([#1107](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1107)) ([89394db](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/89394dbf6ca885abdf0d96627ecc66b7a5fed37b))
* **GroupBrackets:** Don't draw if only one instrument visible ([b72ef4e](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/b72ef4e04f77c0cd9dc2d6a70f9166bed39630d5))
* **Positioning:** Fix ghost notes only created for first few notes in measure, fixing cross stave positioning ([#1062](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1062)) ([0507917](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/05079175993c3a86b35ec15c1fcabe08caa0e4f1))


### Features

* **ChordSymbols:** add EngravingRules.DefaultColorChordSymbol ([#1106](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1106)) ([7f00a9b](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/7f00a9b29eee4df85741bdc4554d7064b4f8fe25))
* **Note:** Store TransposedPitch (for API access) ([72633da](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/72633da238a18c0f41c1a82e24fba26955ba16cf))
* **Slurs:** GraphicalSlurs save their SVGElement (but not GraphicalTies) ([5686daa](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/5686daa7a78ad0dd4058131297e14f0e45988a86))
* **Transposing:** Able to transpose single instrument ([#1115](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1115)) independently of other instruments ([e997df9](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/commit/e997df9464a78c146a7dc20d39be706bb4814e32)), closes [#1116](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1116)

**Comment by sschmidTU** [comment]
Turns out the npm test error in Windows comes from webpack, not directly from karma-webpack,
and it looks like it's not simple to fix:
webpack/webpack#12759

So we'll have to give up windows multiple drive support for now and just merge #1119,
this is the easiest option,
as it works on MacOS (with patch) and Linux now.

**Comment by sschmidTU** [comment]
Webpack server hot reload for changes in `node_modules/` is broken as well...

**Comment by rvilarl** [comment]
@sschmidTU where should I find my `webpack.local.js` ?

**Comment by sschmidTU** [comment]
@rvilarl i don't know, you added it with a npm start:local command in this PR. we can take it out again if it's not necessary.
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/blob/develop/package.json

**Comment by rvilarl** [comment]
@sschmidTU I do not think that I introduced that. It was already there in version 0.9.0

**Pull request merged**